File: buzzcast-telegram-recorder/android-app/app/src/main/java/com/buzzcast/recorder/FacecastApi.java
package com.buzzcast.recorder;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.net.URLEncoder;

import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class FacecastApi {

    private static final String BASE_URL = "https://dhcxzil.facecast.xyz/faceshow";
    private static final String LIVE_URL = "https://live.facecast.xyz/live";

    private final OkHttpClient client;
    private final String userId;
    private final String token;

    public FacecastApi(String userId, String token) {
        this.userId = userId;
        this.token = token;
        this.client = new OkHttpClient.Builder()
                .connectTimeout(15, java.util.concurrent.TimeUnit.SECONDS)
                .readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
                .build();
    }

    private String randomSlashes() {
        int count = (int) (Math.random() * 10);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < count; i++) sb.append('/');
        return sb.toString();
    }

    public JSONObject getUserInfo(String targetUserId) throws IOException {
        String url = BASE_URL + "/tokens/PersonalHome/" + randomSlashes()
                + "findHomeUserInfo?userId=" + targetUserId;

        RequestBody body = new FormBody.Builder()
                .add("systoken", token)
                .add("userId", targetUserId)
                .build();

        Request request = new Request.Builder().url(url).post(body).build();

        try (Response response = client.newCall(request).execute()) {
            if (response.body() != null) {
                return new JSONObject(response.body().string());
            }
        } catch (Exception e) {
            throw new IOException("getUserInfo failed: " + e.getMessage());
        }
        return null;
    }

    public JSONObject getLiveInfo(String targetUserId) throws IOException {
        String url = BASE_URL + "/tokens/ranking/v2/getLastLiveInfoByUserId?liveUserId=" + targetUserId;

        Request request = new Request.Builder().url(url).get().build();

        try (Response response = client.newCall(request).execute()) {
            if (response.body() != null) {
                return new JSONObject(response.body().string());
            }
        } catch (Exception e) {
            throw new IOException("getLiveInfo failed: " + e.getMessage());
        }
        return null;
    }

    public JSONObject getFavorites(int page) throws IOException {
        String url = BASE_URL + "/user/attention/" + randomSlashes() + "listAttention";

        RequestBody body = new FormBody.Builder()
                .add("systoken", token)
                .add("userId", userId)
                .add("currPage", String.valueOf(page))
                .add("pageSize", "20")
                .add("types", "0")
                .build();

        Request request = new Request.Builder().url(url).post(body).build();

        try (Response response = client.newCall(request).execute()) {
            if (response.body() != null) {
                return new JSONObject(response.body().string());
            }
        } catch (Exception e) {
            throw new IOException("getFavorites failed: " + e.getMessage());
        }
        return null;
    }

    public boolean isLive(JSONObject userInfo) {
        try {
            JSONObject result = userInfo.getJSONObject("result");
            return result.getInt("isLive") == 1;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean isPrivateStream(JSONObject userInfo) {
        try {
            JSONObject result = userInfo.getJSONObject("result");
            JSONObject liveData = result.getJSONObject("liveData");
            String flvUrl = liveData.optString("flv_url", "");
            int isLive = result.getInt("isLive");
            return !flvUrl.isEmpty() && isLive == 0;
        } catch (Exception e) {
            return false;
        }
    }

    public String getNickName(JSONObject userInfo) {
        try {
            return userInfo.getJSONObject("result").getString("nickName");
        } catch (Exception e) {
            return "unknown";
        }
    }

    public String getStreamId(JSONObject liveInfo) {
        try {
            return liveInfo.getJSONObject("result").getString("streamId");
        } catch (Exception e) {
            return null;
        }
    }

    public String getLiveId(JSONObject liveInfo) {
        try {
            return liveInfo.getJSONObject("result").getString("liveId");
        } catch (Exception e) {
            return null;
        }
    }

    public String getFlvUrl(String streamId) {
        return LIVE_URL + "/" + streamId + ".flv";
    }

    public String getHlsUrl(String streamId) {
        return LIVE_URL + "/" + streamId + ".m3u8";
    }
}

File: buzzcast-telegram-recorder/facecast-api.js
const axios = require('axios');

const BASE_URL = 'https://dhcxzil.facecast.xyz/faceshow';
const LIVE_URL = 'https://live.facecast.xyz/live';

function randomSlashes() {
  return '/'.repeat(Math.floor(10 * Math.random()));
}

class FacecastAPI {
  constructor(userId, token) {
    this.userId = userId;
    this.token = token;
  }

  // Login avec email + mot de passe -> retourne { userId, token }
  static async login(email, password) {
    const url = `${BASE_URL}/api/sys/login/v2`;
    const res = await axios.post(url, null, {
      params: {
        deviceId: '214035648725148',
        email,
        pwd: password,
        type: 2
      }
    });
    const data = res.data;
    if (data.code === '40007' || data.msg !== 'Login successfully') {
      throw new Error(data.msg || 'Login failed');
    }
    return {
      userId: String(data.result.userId),
      token: data.result.token,
      nickName: data.result.nickName || ''
    };
  }

  async getUserInfo(targetUserId) {
    const url = `${BASE_URL}/tokens/PersonalHome/${randomSlashes()}findHomeUserInfo?userId=${targetUserId}`;
    const res = await axios.post(url, null, {
      params: { systoken: this.token, userId: targetUserId }
    });
    return res.data;
  }

  async getLiveInfo(targetUserId) {
    const url = `${BASE_URL}/tokens/ranking/v2/getLastLiveInfoByUserId?liveUserId=${targetUserId}`;
    const res = await axios.get(url);
    return res.data;
  }

  async getStreamInfo(userId, liveId) {
    const param = JSON.stringify({ likeNum: '', liveId, userId });
    const url = `${BASE_URL}/tokens/live/newLive/${randomSlashes()}getLiveInfo?param=${encodeURI(param)}`;
    const res = await axios.post(url, null, {
      params: { systoken: this.token }
    });
    return res.data;
  }

  async getFavorites(page = 1) {
    const url = `${BASE_URL}/user/attention/${randomSlashes()}listAttention`;
    const res = await axios.post(url, null, {
      params: {
        systoken: this.token,
        userId: this.userId,
        currPage: page,
        pageSize: 20,
        types: 0
      }
    });
    return res.data;
  }

  isPrivateStream(userInfo) {
    if (!userInfo || !userInfo.result) return false;
    const { liveData, isLive } = userInfo.result;
    return liveData && liveData.flv_url !== '' && isLive === 0;
  }

  isLive(userInfo) {
    if (!userInfo || !userInfo.result) return false;
    return userInfo.result.isLive === 1;
  }

  getStreamUrls(streamId) {
    return {
      flv: `${LIVE_URL}/${streamId}.flv`,
      hls: `${LIVE_URL}/${streamId}.m3u8`,
      rtmp: `rtmp://live.facecast.xyz/live/${streamId}`
    };
  }
}

module.exports = FacecastAPI;
